gqcp/include/Operator/SecondQuantized/USQTwoElectronOperator.hpp

```diff
@@ -0,0 +1,511 @@
+// This file is part of GQCG-gqcp.
+// 
+// Copyright (C) 2017-2019  the GQCG developers
+// 
+// GQCG-gqcp is free software: you can redistribute it and/or modify
+// it under the terms of the GNU Lesser General Public License as published by
+// the Free Software Foundation, either version 3 of the License, or
+// (at your option) any later version.
+// 
+// GQCG-gqcp is distributed in the hope that it will be useful,
+// but WITHOUT ANY WARRANTY; without even the implied warranty of
+// MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+// GNU Lesser General Public License for more details.
+// 
+// You should have received a copy of the GNU Lesser General Public License
+// along with GQCG-gqcp.  If not, see <http://www.gnu.org/licenses/>.
+// 
+#pragma once
+
+
+#include "Basis/SpinorBasis/JacobiRotationParameters.hpp"
+#include "Basis/TransformationMatrix.hpp"
+#include "Mathematical/Representation/QCRankFourTensor.hpp"
+#include "Operator/SecondQuantized/USQOneElectronOperator.hpp"
+#include "Processing/RDM/OneRDM.hpp"
+#include "Processing/RDM/TwoRDM.hpp"
+#include "Utilities/miscellaneous.hpp"
+
+#include <array>
+
+
+namespace GQCP {
+
+
+/**
+A class that represents an 'unrestricted second-quantized two-electron operator' suitable for the projection of the non-relativistic Hamiltonian onto an unrestricted spinor basis. It holds the tensor representation of its parameters for both spin components and both mixed spin components, which are (usually) integrals over first-quantized operators
+ * 
+ *  @tparam _Scalar             the scalar type, i.e. the scalar representation of one of the parameters
+ *  @tparam _Components         the number of components of the second-quantized operator
+ */
+template <typename _Scalar, size_t _Components>
+class USQTwoElectronOperator {
+public:
+
+    using Scalar = _Scalar;
+    static constexpr auto Components = _Components;
+
+
+private:
+    std::array<QCRankFourTensor<Scalar>, Components> gs_alphaAlpha;  // all the tensor representations (hence the 's') of the alpha-alpha parameters (integrals) of the different components of this second-quantized operator
+    std::array<QCRankFourTensor<Scalar>, Components> gs_alphaBeta;  // all the tensor representations (hence the 's') of the alpha-beta parameters (integrals) of the different components of this second-quantized operator
+    std::array<QCRankFourTensor<Scalar>, Components> gs_betaAlpha;  // all the tensor representations (hence the 's') of the beta-alpha parameters (integrals) of the different components of this second-quantized operator
+    std::array<QCRankFourTensor<Scalar>, Components> gs_betaBeta;  // all the tensor representations (hence the 's') of the beta-beta parameters (integrals) of the different components of this second-quantized operator
```

Comment:
I would be in favor of calling the parameters and all variables in the likes of `gs_aa`, `gs_bb`, etc. The `_a` and `_b` subscripts are familiar enough for the readers to understand that they are related to alpha and beta.

What do you think?

In any case, we use `under_scores_for_variables` instead of the `lowerCamelCase` that you've used, but no problem!

Reply:
I've written it this way in order to maintain the same style as the USQOneElectronOperator, where the "alpha" and "beta" subscripts are written completely. I don't mind changing it however, i'll adapt to whatever style is more in line with the rest of the code base. 

If you'd prefer to change it to _aa/_ab/_ba/_bb, I suggest changing it as well in the USQOneElectronOperator, just to maintain consistency. 

The underscore vs. camelcase issue will be addressed.

Reply:
That would be great!